### ISSUE TITLE
ViteJS - Uncaught Error: useRoutes() may be used only in the context of a <Router> component

### ISSUE THREAD

**Issue by bhaskarmurthy** [marmelab/react-admin#7937]
**What you were expecting:**
react-admin should render as expected

**What happened instead:**
react-router does not render any routes, and instead throws an exception:
```
Uncaught Error: useRoutes() may be used only in the context of a <Router> component
```

**Steps to reproduce:**
1. Set up a new Vite project using `react-ts` template
2. Follow tutorial to set up a react-admin component
3. Launch app

**Related code:**

* Preferably, a CodeSandbox forked from 
  - https://codesandbox.io/s/github/marmelab/react-admin/tree/master/examples/simple (v4)
  - https://codesandbox.io/s/github/marmelab/react-admin/tree/3.x/examples/simple (v3)
* A link to a GitHub repo with the minimal codebase to reproduce the issue

```
insert short code snippets here
```

**Other information:**

**Environment**

* React-admin version:
* Last version that did not exhibit the issue (if applicable):
* React version:
* Browser:
* Stack trace (in case of a JS error):
```
Uncaught Error: useRoutes() may be used only in the context of a <Router> component.
    invariant index.tsx:24
    useRoutes index.tsx:642
    Routes index.tsx:344
    React 12
    unstable_runWithPriority scheduler.development.js:468
    React 4
    unstable_runWithPriority scheduler.development.js:468
    React 3
    workLoop scheduler.development.js:417
    flushWork scheduler.development.js:390
    performWorkUntilDeadline scheduler.development.js:157
    js scheduler.development.js:180
    js scheduler.development.js:645
    __require chunk-CEDYW5NB.js:11
    js index.js:6
    __require chunk-CEDYW5NB.js:11
    React 2
    __require chunk-CEDYW5NB.js:11
    js React
    __require chunk-CEDYW5NB.js:11
    <anonymous> react-dom:1
index.tsx:24:19
```

**Comment by septentrion-730n** [comment]
Thanks for your report. We'lll take a look at this.

**Comment by septentrion-730n** [comment]
I went through the first steps of the tutorial without any trouble.
Here is a small [test repository](https://github.com/septentrion-730n/rao-issue7937), please give it a try.

**Comment by bhaskarmurthy** [comment]
Thanks @septentrion-730n . Will investigate a bit more on my end. Appreciate the quick turnaround!

**Comment by MBach** [comment]
I'm starting to have this bug with `react-router-dom: 6.4.0`

**Comment by fzaninotto** [comment]
@MBach Please provide a reproduction if you think this is a react-admin bug.

**Comment by MBach** [comment]
> @MBach Please provide a reproduction if you think this is a react-admin bug.

Well, currently I'm short in time, like most of us, but I'll try to create a repro.

**Comment by crates** [comment]
I'm also having a very similar bug to this with react-router-dom 6.4.1 - if you need to reproduce, just upgrade any of your own React Admin projects (that use useRoutes or useLocation) to the same version of React Router DOM, and it should occur.

I'm also getting this error after upgrading React Router DOM:
`[Route] is not a <Route> component. All component children of <Routes> must be a <Route> or <React.Fragment>`

**Comment by fzaninotto** [comment]
Confirmed: react-router 6.4.1 breaks react-admin. Reopening and marking this as a bug.

**Comment by crates** [comment]
@fzaninotto -

I hear you. I tried reproducing this error with a fresh clone of one of the repositories above. I couldn't get the bug to manifest in the repro, and that does make it seem like the issue isn't with React Admin. But, I've been maintaining our React Admin project for about 2 years now, and we've never had this problem before. It's a fairly sophisticated project with a decently-large scope. (I've been in talks with our company for a couple of months to get us an enterprise license as well, by the way... it's just that we're a huge company and things move really slow.)

The thing is, we didn't have any problems like this for the last couple of years we've been using React Admin. We didn't have these problems when we finished the upgrade from version 3 to version 4 of React Admin (and upgraded React Router to version 6, as well). The only time we have these unique errors is when we upgrade from version 6.3.0 of React Router to version 6.4.x, and the errors are coming from somewhere inside React Admin. It's something to do with the context from React Router not making it through into the Admin component. When we downgraded back to 6.3.0 of React Router, all of these strange bugs went away again.

Moreover, when I run `npm ls`, I don't see any competing versions of React Router - just whichever one is in our package.json at the time (6.3.0 when it's working, and 6.4.1 when it's broken). And what's more, I'm the third person on this ticket reporting basically identical issues... so it seems a bit dismissive to assume we're all doing this wrong. Our projects worked fine until we upgraded React Router. Is it possible there's some kind of important difference (or bug) in the latest version that hasn't yet been addressed in React Admin?

**Comment by crates** [comment]
@MBach / @bhaskarmurthy -

Boy, do I feel foolish. @fzaninotto was 100% correct: this bug had nothing to do with React Admin (at least, for me). Up until React Router v6.4.x it was possible to import the Route component from `react-router` without any breaking issues. As of the latest version, however, you need to correctly import it from `react-router-dom` instead. In fact, this has been the correct way to do things for a while now. It's further complicated and confusing because it's still possible to import it from `react-router` without any TypeScript or linting errors, but that's what was causing the problems on our end.

Can you both check to see if you've made the same mistake? If so, I hope this solves your problems as well!

**Comment by crates** [comment]
So, once more for clarity, do NOT do this if you have any custom routes in your app:
```
import {
  Route,
} from 'react-router';
```

Instead, this is the correct approach which will hopefully fix your issues:
```
import {
  Route,
} from 'react-router-dom';
```

God, I really do _hate_ React Router and its endless parade of syntax changes.

**Comment by MBach** [comment]
> So, once more for clarity, do NOT do this if you have any custom routes in your app:
> 
> ```
> import {
>   Route,
> } from 'react-router';
> ```
> 
> Instead, this is the correct approach which will hopefully fix your issues:
> 
> ```
> import {
>   Route,
> } from 'react-router-dom';
> ```
> 
> God, I really do _hate_ React Router and its endless parade of syntax changes.

Good asumption with custom routes, that's what I thought too. However, I do have checked my code, and I have react-router-dom only.

**Comment by cwagner22** [comment]
@MBach Did you solve it?

**Comment by qpointsystems** [comment]
Another anecdote related to this root issue that might be helpful.

Was just doing an upgrade pass on my application, going through my dependencies. As part of that process I upgraded to **react-router-dom** 6.6.1 from 6.4.2 and encountered the issue above.  After hunting around for a possible cause, I came across this thread here in GitHub.   

I played around with some trial and error,  upgrading to 6.4.5, no go.  6.5.0, no go.  

Hmmm.  Then, removed it entirely doing some cleanup, was checking something else out, and the browser page refresh and the app did not through the error.  Crazy, haha.

To summarize my situation: I previously had  "react-router-dom": "^6.x.x" explicitly added to my package.json.  
Removing it entirely seems to have solved the problem.

**Comment by openforce-jp** [comment]
I had the same issue using react-router@6.4.4.
After that, changing to 6.3.0 didn't solve the problem.
I'm using **pnpm** and had two versions of react-router in /node_modules/.pnpm/.
I solved it by deleting 6.4.4 from the directory inside  /node_modules/.pnpm/.

**Comment by RedRush1** [comment]
I also encountered this error. What is interesting is that it gets reproduced when react-router-dom gets new version. To reproduce it locally I delete react-router-dom and install again. After that react-admin break consistently when you click on resource. Temporal fix is to update react-router-dom to latest version, however after third time it is rather unpleasant.

**Comment by CLSnazel** [comment]
Chiming in for anyone else googling this issue...

We encountered this error while moving react-admin from one framework to a fresh Create React App. For this case, we wanted to avoid deleting `yarn.lock` so we didn't get any further dependency changes. Adding react-router-dom and react-router to the resolutions in `package.json` removed the duplicate versions:

```json
"resolutions":{
  "react-router-dom": "6.8.0",
  "react-router": "6.8.0"
}
```

**Comment by ythomop** [comment]
Just encountered this error and I solved it by using react-admin: "4.7.4".
Previously, I had "^4.7.4", which installed version 4.7.6 which seems to be broken, for me at least.

**Comment by teodora-zhelyazkova** [comment]
I just had this issue and solved it by deleting `node_modules ` and `pnpm-lock.yaml` & running `pnpm install` again

**Comment by psam44** [comment]
I confirm duplicates as the cause of the problem.
After some confused add/remove commands and manual edit of package.json, I ended up with a mess in yarn.lock.
Extracts:
```
react-admin@^4.16.6:
  version "4.16.6"
  dependencies:
    react-router "^6.1.0"
    react-router-dom "^6.1.0"
react-router-dom@^6.1.0:
  version "6.3.0"
  dependencies:
    history "^5.2.0"
    react-router "6.3.0"
react-router-dom@^6.21.2:
  version "6.21.2"
  dependencies:
    "@remix-run/router" "1.14.2"
    react-router "6.21.2"
react-router@6.21.2:
  version "6.21.2"
  dependencies:
    "@remix-run/router" "1.14.2"
react-router@6.3.0, react-router@^6.1.0:
  version "6.3.0"
  dependencies:
    history "^5.2.0"
```
After manual cleanup of yarn.lock to keep only references to 6.21.2, the server is fine again.

**Comment by splatch** [comment]
Error message is kind of obvious, but I was blinded by similarity of elements. I found this issue by looking for same error message. The root cause for me was that I missed `<Router>` element in the app. I had `Routes` and `Route` parts, but by accident removed its parent while juggling stuff.  Once I brought back `Router` (aka BrowserRouter) everything started to work again.

**Comment by pefad** [comment]
who have solved this error, am encountering the same here:Uncaught Error: useRoutes() may be used only in the context of a <Router> component.

**Comment by rvanlaak** [comment]
Landing here, after looking for router related issues. Getting the same type of error. Tried pinning it down to the specific dependency, but could not manage to find it.

For me the error specifically was not pointing toward a specific component or page so that made it quite hard to debug:

```cli
Error: useLocation() may be used only in the context of a <Router> component.
    at invariant (https://localhost/_next/static/chunks/node_modules__pnpm_b5787fba._.js:1774:15)
    at useLocation (https://localhost/_next/static/chunks/node_modules__pnpm_b5787fba._.js:5790:5)
    at useResetErrorBoundaryOnLocationChange (https://localhost/_next/static/chunks/386ad_ra-core_dist_esm_a4febb8d._.js:206:345)
    at Error (https://localhost/_next/static/chunks/8bd6f_%40api-platform_admin_lib_b0bd0dca._.js:3174:449)
    at react-stack-bottom-frame (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:13596:24)
    at renderWithHooks (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:3560:24)
    at updateFunctionComponent (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:5317:21)
    at beginWork (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:5916:24)
    at runWithFiberInDEV (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:1335:74)
    at performUnitOfWork (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:7988:97)
    at workLoopSync (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:7880:40)
    at renderRootSync (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:7863:13)
    at performWorkOnRoot (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:7602:212)
    at performWorkOnRootViaSchedulerTask (https://localhost/_next/static/chunks/a14e7_react-dom_638ad3bb._.js:8566:9)
    at MessagePort.performWorkUntilDeadline (https://localhost/_next/static/chunks/node_modules__pnpm_a301e626._.js:1339:64)
```

The logs I get when running `npm ls react-router`:
```cli
root@b934c0e5f00c:/srv/app# npm ls react-router
pwa@dev-86bf2fa /srv/app
+-- @api-platform/admin@4.0.6 -> ./node_modules/.pnpm/@api-platform+admin@4.0.6_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3__4529acb264bf9be0fa00efeb0b8e5746/node_modules/@api-platform/admin
| `-- react-admin@5.8.1 -> ./node_modules/.pnpm/react-admin@5.8.1_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3_react@19_44c76ad40c0cd0b375669a6a8c31fe3f/node_modules/react-admin
|   `-- react-router@7.6.0 deduped invalid: "^6.28.1" from node_modules/.pnpm/react-admin@5.8.1_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3_react@19_44c76ad40c0cd0b375669a6a8c31fe3f/node_modules/react-admin, "^6.28.1" from node_modules/.pnpm/react-admin@5.8.1_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3_react@19_44c76ad40c0cd0b375669a6a8c31fe3f/node_modules/react-admin -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
+-- ra-core@5.8.0 -> ./node_modules/.pnpm/ra-core@5.8.0_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_074bcc19d154297a0e07ca00f906b4fd/node_modules/ra-core
| +-- react-router-dom@6.30.0 -> ./node_modules/.pnpm/react-router-dom@6.30.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router-dom
| | `-- react-router@6.30.0 -> ./node_modules/.pnpm/react-router@6.30.0_react@19.1.0/node_modules/react-router
| |   `-- react-router-dom@6.30.0 -> ./node_modules/.pnpm/react-router-dom@6.30.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router-dom
| |     `-- react-router@6.30.0 deduped -> ./node_modules/.pnpm/react-router@6.30.0_react@19.1.0/node_modules/react-router
| `-- react-router@7.6.0 invalid: "^6.28.1" from node_modules/.pnpm/ra-core@5.8.0_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_074bcc19d154297a0e07ca00f906b4fd/node_modules/ra-core -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
+-- react-admin@5.8.1 -> ./node_modules/.pnpm/react-admin@5.8.1_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3_react@19_44c76ad40c0cd0b375669a6a8c31fe3f/node_modules/react-admin
| +-- ra-core@5.8.1 -> ./node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core
| | +-- react-router-dom@6.30.0 -> ./node_modules/.pnpm/react-router-dom@6.30.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router-dom
| | | `-- react-router@6.30.0 deduped -> ./node_modules/.pnpm/react-router@6.30.0_react@19.1.0/node_modules/react-router
| | `-- react-router@7.6.0 invalid: "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
| +-- ra-i18n-polyglot@5.8.1 -> ./node_modules/.pnpm/ra-i18n-polyglot@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1_bc75b94e00ec30b329b7a636cefea66c/node_modules/ra-i18n-polyglot
| | `-- ra-core@5.8.1 -> ./node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core
| |   `-- react-router@7.6.0 deduped invalid: "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core, "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
| +-- ra-language-english@5.8.1 -> ./node_modules/.pnpm/ra-language-english@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@1_bd23b6be2654682b7670f4755d6410c7/node_modules/ra-language-english
| | `-- ra-core@5.8.1 -> ./node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core
| |   `-- react-router@7.6.0 deduped invalid: "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core, "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core, "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
| +-- ra-ui-materialui@5.8.1 -> ./node_modules/.pnpm/ra-ui-materialui@5.8.1_b40de75f70f89d3a0af00d2326f4185f/node_modules/ra-ui-materialui
| | +-- ra-core@5.8.1 -> ./node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core
| | | `-- react-router@7.6.0 deduped invalid: "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core, "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core, "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core, "^6.28.1" from node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/ra-core -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
| | +-- react-router-dom@6.30.0 -> ./node_modules/.pnpm/react-router-dom@6.30.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router-dom
| | | `-- react-router@6.30.0 deduped -> ./node_modules/.pnpm/react-router@6.30.0_react@19.1.0/node_modules/react-router
| | `-- react-router@7.6.0 invalid: "^6.28.1" from node_modules/.pnpm/ra-ui-materialui@5.8.1_b40de75f70f89d3a0af00d2326f4185f/node_modules/ra-ui-materialui -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
| +-- react-router-dom@6.30.0 -> ./node_modules/.pnpm/react-router-dom@6.30.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router-dom
| | `-- react-router@6.30.0 deduped -> ./node_modules/.pnpm/react-router@6.30.0_react@19.1.0/node_modules/react-router
| `-- react-router@7.6.0 invalid: "^6.28.1" from node_modules/.pnpm/react-admin@5.8.1_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3_react@19_44c76ad40c0cd0b375669a6a8c31fe3f/node_modules/react-admin, "^6.28.1" from node_modules/.pnpm/react-admin@5.8.1_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3_react@19_44c76ad40c0cd0b375669a6a8c31fe3f/node_modules/react-admin -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router
`-- react-router@7.6.0 -> ./node_modules/.pnpm/react-router@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router

npm error code ELSPROBLEMS
npm error invalid: react-router@7.6.0 /srv/app/node_modules/.pnpm/ra-core@5.8.0_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_074bcc19d154297a0e07ca00f906b4fd/node_modules/react-router
npm error invalid: react-router@7.6.0 /srv/app/node_modules/.pnpm/react-admin@5.8.1_@mui+system@7.1.0_@emotion+react@11.14.0_@types+react@19.1.3_react@19_44c76ad40c0cd0b375669a6a8c31fe3f/node_modules/react-router
npm error invalid: react-router@7.6.0 /srv/app/node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.1_react@19.1.0__react_2f19e4d29fcb5b909784080525ece3a8/node_modules/react-router
npm error invalid: react-router@7.6.0 /srv/app/node_modules/.pnpm/ra-ui-materialui@5.8.1_b40de75f70f89d3a0af00d2326f4185f/node_modules/react-router
npm error A complete log of this run can be found in: /root/.npm/_logs/2025-05-12T14_48_59_203Z-debug-0.log
root@b934c0e5f00c:/srv/app# 
```

So the conflict does looks related to `ra-core`, `ra-ui-materialui`, `react-admin`? Same happens when trying this with the `react-router-dom` component, and when using `v6.30` of that.

Had to wrap the entire `<HydraAdmin />` component in a `<MemoryRouter />` to eventually isolate the entire admin to solve this:

```tsx
import {MemoryRouter} from "react-router";

const App = () => (
  <MemoryRouter>
    <HydraAdmin
      entrypoint={ENTRYPOINT}
      ...
```

**Comment by slax57** [comment]
@rvanlaak All 3 RA packages declare the dependency (or peer dependency respectively) on `react-router` as they should:

https://github.com/marmelab/react-admin/blob/391b4d7715917405f57a8607a71dbc6a8fd3ecb4/packages/react-admin/package.json#L49-L50

https://github.com/marmelab/react-admin/blob/391b4d7715917405f57a8607a71dbc6a8fd3ecb4/packages/ra-ui-materialui/package.json#L67-L68

https://github.com/marmelab/react-admin/blob/391b4d7715917405f57a8607a71dbc6a8fd3ecb4/packages/ra-core/package.json#L56-L57

It seems that, for some reason, **pnpm** is still creating duplicates although `7.6.0` is a valid version matching those ranges.

Can you try using [pnpm overrides](https://www.squash.io/how-to-use-pnpm-overrides-for-dependency-management/) to see if that solves the issue?

```json
// package.json
"pnpm": {
  "overrides": {
    "react-router": "^7.6.0",
    "react-router-dom": "^7.6.0"
  }
}
```

**Comment by rvanlaak** [comment]
@slax57 tested that approach, it results in a build log entry:

> ⨯ Error: Failed to load external module @api-platform/admin: Error [ERR_REQUIRE_CYCLE_MODULE]: Cannot require() ES Module /srv/app/node_modules/.pnpm/react-router-dom@7.6.0_react-dom@19.1.0_react@19.1.0__react@19.1.0/node_modules/react-router-dom/dist/index.mjs in a cycle. (from /srv/app/node_modules/.pnpm/ra-core@5.8.1_react-dom@19.1.0_react@19.1.0__react-hook-form@7.56.3_react@19.1.0__react_67202adf6936d8951d78a9713cf443c2/node_modules/ra-core/dist/cjs/auth/useLogout.js)

**Comment by tberriot** [comment]
> [@rvanlaak](https://github.com/rvanlaak) All 3 RA packages declare the dependency (or peer dependency respectively) on `react-router` as they should:
> 
> [react-admin/packages/react-admin/package.json](https://github.com/marmelab/react-admin/blob/391b4d7715917405f57a8607a71dbc6a8fd3ecb4/packages/react-admin/package.json#L49-L50)
> 
> Lines 49 to 50 in [391b4d7](/marmelab/react-admin/commit/391b4d7715917405f57a8607a71dbc6a8fd3ecb4)
> 
>  "react-router": "^6.28.1 || ^7.1.1", 
>  "react-router-dom": "^6.28.1 || ^7.1.1" 
> [react-admin/packages/ra-ui-materialui/package.json](https://github.com/marmelab/react-admin/blob/391b4d7715917405f57a8607a71dbc6a8fd3ecb4/packages/ra-ui-materialui/package.json#L67-L68)
> 
> Lines 67 to 68 in [391b4d7](/marmelab/react-admin/commit/391b4d7715917405f57a8607a71dbc6a8fd3ecb4)
> 
>  "react-router": "^6.28.1 || ^7.1.1", 
>  "react-router-dom": "^6.28.1 || ^7.1.1" 
> [react-admin/packages/ra-core/package.json](https://github.com/marmelab/react-admin/blob/391b4d7715917405f57a8607a71dbc6a8fd3ecb4/packages/ra-core/package.json#L56-L57)
> 
> Lines 56 to 57 in [391b4d7](/marmelab/react-admin/commit/391b4d7715917405f57a8607a71dbc6a8fd3ecb4)
> 
>  "react-router": "^6.28.1 || ^7.1.1", 
>  "react-router-dom": "^6.28.1 || ^7.1.1" 
> It seems that, for some reason, **pnpm** is still creating duplicates although `7.6.0` is a valid version matching those ranges.
> 
> Can you try using [pnpm overrides](https://www.squash.io/how-to-use-pnpm-overrides-for-dependency-management/) to see if that solves the issue?
> 
> // package.json
> "pnpm": {
>   "overrides": {
>     "react-router": "^7.6.0",
>     "react-router-dom": "^7.6.0"
>   }
> }

Hi @slax57 
Thanks a lot for this fix !
I've just been install RA with `npm create react-admin@latest` and `package.json` does not include this fix. Maybe it could be usefull to add it for next users could save time ?

**Comment by tberriot** [comment]
Thanks @slax57 
I use `yarn v1.22.22` to install packages.
I think I found the problem : I've launch create-react-admin with npx because installation with yarn didn't worked, so CRA doesn't add this line for yarn as packageManager.
I will add this line manually. 👍
